Add DINOv3ViTForImageClassification support

dimidagd opened the pull request:
@yonigozlan @molbap

molbap reviewed (commented):
Thanks! Left two comments

molbap commented on src/transformers/models/dinov3_vit/modeling_dinov3_vit.py:
```diff
@@ -530,4 +530,64 @@ def forward(
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.dinov3.embeddings.patch_embeddings
```
This should not be needed now
```suggestion
```

dimidagd replied:
I had a test failing before adding it actually, how come is not needed?

molbap reviewed (approved):
Sure, just need to check `get_input_embeddings`, as in most cases it's not needed to add it and we prefer to add as little code as possible. It's not needed in Dinov2, why here? 
Otherwise LGTM and ok to merge once this is sorted out!

dimidagd commented:
@molbap what is the process for merging PRs after they have been approved?  Trying to understand better the contribution workflow

Best 
DD

molbap reviewed (commented):
I have a pending question on which test was failing for the embeddings, please let me know! apart from that for the review process when reviewers have approved we ping the core maintainers to approve and merge your PR!

This one in particular launched an internal discussion since we're trying to minimize the maintenance surface, see here https://github.com/huggingface/transformers/pull/41450 but it'll wait a follow-up PR.

molbap reviewed (commented):
Added a missed `head_mask`, other tham that ping me when you know for the failing test and we can merge this! With https://github.com/huggingface/transformers/pull/41276 it's nice additions to dinov3

dimidagd commented:
> Added a missed `head_mask`, other tham that ping me when you know for the failing test and we can merge this! With #41276 it's nice additions to dinov3

Here is the failing test

https://app.circleci.com/jobs/github/huggingface/transformers/1982122

dimidagd commented:
> Added a missed `head_mask`, other tham that ping me when you know for the failing test and we can merge this! With #41276 it's nice additions to dinov3

@molbap upon rereading your comments, I think you implied that adding the get_input_embeddings should not be necessary. 

Here is the failing test, which was copied over from dinov2.

https://app.circleci.com/jobs/github/huggingface/transformers/1993707

dimidagd commented:
I will see if I can find the time for this, what are your thoughts on the
`get_input_embeddings` test?

On Fri, 31 Oct 2025 at 16:03, Pablo Montalvo ***@***.***>
wrote:

> *molbap* left a comment (huggingface/transformers#41224)
> <https://github.com/huggingface/transformers/pull/41224#issuecomment-3473483964>
>
> Hi @dimidagd <https://github.com/dimidagd> , circling back to this, this
> is to support the classification head released by Meta for dinov3, right?
> In that case, it would also be needed to have a small test to load that
> head and test it works to avoid future regressions. I'm referring to this
> https://github.com/facebookresearch/dinov3?tab=readme-ov-file#pretrained-heads---image-classification
>
> That way there'd be more ground to support the implementation of a new
> head!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/transformers/pull/41224#issuecomment-3473483964>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALECAUJ5KWWT4LDURHPDCB332N24LAVCNFSM6AAAAACH4V2IVCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTINZTGQ4DGOJWGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
[image: photo]
*Dagdilelis Dimitris | PhD*
m:+45 52 60 18 25 <+45+52+60+18+25> | ***@***.*** | a:Gullandsgade
18, 2800 Lyngby
<http://us.linkedin.com/in/dimidagd>

dimidagd commented:
> @dimidagd can you use a Colab? 👀 using small checkpoints to validate implementation is ok!

@merveenoyan hi again, meta released the classification head only for the 7B model. I can't fit that on colab resources either (GPU or CPU). I can't imagine myself  implementing such tests from a colab env tbh.

dimidagd commented:
@molbap I uploaded the 7b weights with the linear classifier adapter on HF, and wrote a simple test on the cat COCO sample in the repo. Perhaps you could run the slow tests?

dimidagd committed:
Add Dinov3ViT support for image-classification task

dimidagd committed:
update auto-modeling configuration

dimidagd committed:
write tests

- Slow test checking logits and predicted class based on COCO cat sample
- Migrated tests from Dinov2

dimidagd committed:
docs: Update docs to reflect support for classification task

dimidagd committed:
The DINOv3 backbone must live in `self.dinov3_vit` rather than `self.dinov3`.
When the backbone is stored under the wrong attribute, both `AutoModelForImageClassification.from_pretrained` and `DINOv3ViTForImageClassification` created from a headless checkpoint fail to load weights correctly because the state dict cannot map the backbone parameters.

This change aligns the class with the expected naming used by the loader utilities and restores correct weight loading behavior.

dimidagd committed:
Uses `@require_torch_large_accelerator` to run test with high VRAM requirement

dimidagd commented on tests/models/dinov3_vit/test_modeling_dinov3_vit.py:
```diff
@@ -274,6 +304,33 @@ def default_image_processor(self):
         )
 
+    @slow
+    def test_inference_lc_head_imagenet(self):
```
Resolved in bcdd666873bd1946d474390799e423d2523f3501

dimidagd committed:
fix: moves tests to cpu because not enough VRAM avail

dimidagd commented:
@molbap pinging you to hear more abt this PR :)

molbap commented on tests/models/dinov3_vit/test_modeling_dinov3_vit.py:
```diff
@@ -238,6 +269,12 @@ def test_model_from_pretrained(self):
         self.assertIsNotNone(model)
 
+    @slow
+    def test_model_for_image_classification_from_pretrained(self):
```
will also require the large accelerator no? Let's remove it, it's tested in the Integration test below

molbap reviewed (commented):
Hello! Now that this is merged with main, some things have changed a bit in terms of code quality. The rest is ok

dimidagd committed:
Add image classification usage snippet to DINOv3 docs

dimidagd committed:
style

dimidagd committed:
fix: modular

dimidagd committed:
Merge branch 'main' into feat/dinov3forimageclassification
